auth/src/main/java/com/farsitel/bazaar/auth/connection/ServiceAuthConnection.kt

```diff
@@ -0,0 +1,136 @@
+
+    private var authService: InAppBazaarAuth? = null
+    private var connectionThreadSecure: AbortableCountDownLatch? = null
+
+    override fun getLastAccountId(
+        owner: LifecycleOwner?,
+        callback: BazaarSignInCallback
+    ) {
+        withServiceInBackground(
+            func = { connection ->
+                val response = getLastAccountBundleFromService(connection)
+                callback.onAccountReceived(response)
+            }, onException = {
+                callback.onAccountReceived(getLastAccountResponse(extras = null))
+            })
+    }
+
+    override fun getLastAccountIdSync(
+        owner: LifecycleOwner?
+    ): BazaarResponse<BazaarSignInAccount> {
+        return withService(
+            func = { connection ->
+                getLastAccountBundleFromService(connection)
+            }, onException = {
+                getLastAccountResponse(extras = null)
+            }
+        )
+    }
+
+    private fun getLastAccountBundleFromService(
+        connection: InAppBazaarAuth
+    ): BazaarResponse<BazaarSignInAccount> {
+        val bundle = connection.getLastAccountId(context.packageName)
+        return getLastAccountResponse(bundle)
+    }
+
+    override fun disconnect(context: Context) {
+        context.unbindService(this)
+
+        authService = null
+        connectionThreadSecure?.abort()
+
+        backgroundThread.interrupt()
+        super.disconnect(context)
+    }
+
+    fun connect(): Boolean {
+        return Intent(AUTH_SERVICE_ACTION).apply {
+            `package` = BAZAAR_PACKAGE_NAME
+        }.takeIf {
+            context.packageManager.queryIntentServices(it, 0).isNullOrEmpty().not()
+        }?.let {
+            try {
+                val connect = context.bindService(it, this, Context.BIND_AUTO_CREATE)
+                if (connect) {
+                    connectionThreadSecure = AbortableCountDownLatch(1)
+                }
+                connect
+            } catch (e: SecurityException) {
+                false
+            }
+        } ?: false
+    }
+
+    override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
+        authService = InAppBazaarAuth.Stub.asInterface(service)
+        connectionThreadSecure?.countDown()
+    }
+
+    override fun onServiceDisconnected(name: ComponentName?) {
+        authService = null
+    }
+
+    private fun <T> withService(
+        func: (InAppBazaarAuth) -> T,
+        onException: () -> T
+    ): T {
+        return try {
+            connectionThreadSecure?.await()
+            if (authService == null) {
+                connect()
+            }
+
+            connectionThreadSecure?.await()
+            func.invoke(requireNotNull(authService))
+        } catch (e: Exception) {
+            onException.invoke()
+        }
+    }
+
+    private fun <T> withServiceInBackground(
+        func: (InAppBazaarAuth) -> T,
+        onException: () -> T
+    ) {
+        val runnable = {
```

Comment:
Move this logic to another method in order to DRY principle.

Reply:
Done